source/event.d

```diff
@@ -1,51 +1,360 @@
 module event;
 
-abstract class Event
+
+/** Main event class
+ *
+ * This is class is abstract and only used to establish some default functions
+ *     to every extended Event. \
+ */
+@safe abstract class Event
 {
 public:
-	abstract @property EventType eventType() const;
-	abstract @property string name() const;
+	@safe pure
+	abstract @property string eventType() const;
+
 
-	pragma(inline, true)
+	pragma(inline, true) @safe pure
 	override string toString() const
 	{
-		return name;
+		return eventType;
 	}
 
-	bool isEventTypeOf(in EventType eType) const
+
+	/** Compare two EventTypes
+	 *
+	 * Params:
+	 *     otherType = type to be compared
+	 *
+	 * Returns:
+	 *     `true` if both types are equal. \
+	 *     `false` otherwise
+	 */
+	@safe pure
+	bool isEventTypeOf(in string otherType) const
 	{
-		return eventType == eType;
+		return eventType == otherType;
 	}
 }
 
-enum EventType
+
+/** Default EventType struct
+ *
+ * Used as the main data storage for the events. \
+ * You must define this as an UDA when declaring your custom events. \
+ *
+ * Examples:
+ * --------------------
+ * @EventType("MyEvent")
+ * class MyEvent : Event { // ... }
+ * --------------------
+ *
+ * Note: To use the `basicEventType` mixin template you need to define an
+ *     EventType UDA otherwise it'll not work! You CAN create your own EventType
+ *     version without UDAs but keep in mind you'll have to implement your custom
+ *     logic of the abstract methods. The EventDispacher is not afected by this.
+ */
+@safe struct EventType
 {
-	None,
-	CreatedEvent,
+	const string name;
 }
 
-mixin template basicEventType(T)
+
+/** Quick way to implement your default abstract and static methods
+ *
+ * Call this mixin template on your custom event. This will generate your getters
+ *     for the EventType.
+ *
+ * Examples:
+ * --------------------
+ * @EventType("MyEvent")
+ * class MyEvent : Event
+ * {
+ *     mixin basciEventType!MyEvent;
+ * }
+ * --------------------
+ *
+ * Note: You need to declare an EventType UDA otherwise this won't compile!
+ *
+ * Params:
+ *     T = class type extended from Event
+ */
+mixin template basicEventType(T : Event)
 {
 public:
-	pragma(inline, true)
```

Comment:
I do understand what you want here. You may got this from C++, but I do thing `pragma(inline, true)` should only be used if you really have to, because you may end up having heavier binaries if you use it in a wrong way.

I do prefer the compiler choice but this is up to you, but be aware of the side effects of this.

Reply:
But the function will always have the same behaviour, the compiler by default does not inline functions unless you tell it to.

Reply:
> But the function will always have the same behaviour, the compiler by default does not inline functions unless you tell it to.

What? No, you are completely wrong. Let me deeply explain what inline does in D and in general what is the side effects of this:

Inlining, in general, should be a thing made by the compiler, in the optimization process.

The D spec says:
> - The default inline behavior is typically selectable with a compiler switch such as `-inline`.
> - Whether a particular function can be inlined or not is implementation defined.

"typically" doesn't mean "always". In practice, the DMD (default D compiler that properly implements D spec) doesn't inline in the optimization process, in general. See, here https://godbolt.org/z/qGMWMo, I ran `dmd` with `-O` for optimized code and also `-release` to suppress some debug behaviors, and even with `pragma(inline, true)` the function is not inlined.

On other hand, LDC compiler (llvm compiler implementation for D), without any flag (default behavior), it does inline when `pragma(inline, true)` is in action ( https://godbolt.org/z/K9EMrc ), although, if you leave the default behavior (removing `pragma(inline, true)`) it doesn't do inlining ( https://godbolt.org/z/vPW6s5 ), unless you tell to optimize it ( https://godbolt.org/z/nTdeaz ).

TL;DR

The compile DOES inline in some cases, unless you force (inline true) or force not to (inline false).

See, here, I don't force inlining and the function is inlined by the compiler on `-O3`: https://godbolt.org/z/q95KG9

Forcing inlining could lead to some side effects like heavy binaries when inlining really big functions when they are called several times and make debug process harder ( see the bottom of this article: http://sourceware.org/gdb/current/onlinedocs/gdb/Inline-Functions.html ).

Conclusion: As I discussed earlier with you, don't do `pragma(inline, true)` or `pragma(inline, false)` unless you really need to or you really know what you are doing.

Reply:
Thank you for this detailed explanation. I thought `prama(inline)` worked diferently. I'll change this in the third PR since this doesn't break anything here.